src/oracle/middleware.js

```diff
@@ -0,0 +1,761 @@
+ * @returns whether last submission was made in less than SEND_CHECK_INTERVAL seconds
+ */
+const checkIfInSubmission = async (feed) => {
+
+  //get last submission time
+  let query = await queryTable("jobs", ["last_submission_time"], feed)
+  //get seconds since last price submission
+  let timePassedSinceSubmission = (Date.now()/1000) - 
+  query.last_submission_time
+  return timePassedSinceSubmission < Number(SEND_CHECK_INTERVAL)
+}
+
+/**
+  * Function to push price on chain to the smart wallet
+  * @param {*} price price to push
+  * @param {*} feed feed to push price to
+  * @param {*} round round to push result to
+  * @param {*} from account to push from
+  * @returns whether successful
+  */
+const pushPrice = async (price, feed, round, from) => {
+
+  //create an offerId with the Date number
+  var offerId = Date.now()
+
+  //get offers
+  let offers = await getOraclesInvitations()
+  //get previous offer for feed
+  let previousOffer = offers[feed]
+
+  //create an offer
+  const offer = {
+    id: Number(offerId),
+    invitationSpec: {
+      source: 'continuing',
+      previousOffer: Number(previousOffer),
+      invitationMakerName: 'PushPrice',
+      invitationArgs: harden([{ unitPrice: BigInt(price), roundId: round }]),
+    },
+    proposal: {},
+  };
+
+  //create keyring
+  var keyring = {
+    "home": "",
+    "backend": "test"
+  }
+
+  //check if submitted for round
+  let submitted = await checkSubmissionForRound(from, previousOffer, round)
+
+  //check if in submission
+  let inSubmission = await checkIfInSubmission(feed)
+
+  //loop retries
+  for (let i = 0; i < SUBMIT_RETRIES && !submitted && !inSubmission; i++) {
+
+    //query round
+    let latestRound = await queryRound(feed)
+    
+    //if latestRound is greater than round being pushed or submission to the round is already made, abort
+    if (latestRound.round_id > round || (latestRound.round_id == round && latestRound.submission_made)){
+      console.log("Price failed to be submitted for old round", round)
+      return false
+    }
+
+    console.log("Submitting price for round", round, "try", (i + 1))
+
+    offer.id = Number(Date.now())
+
+    //output action
+    var data = outputAction({
+      method: 'executeOffer',
+      // @ts-ignore
+      offer,
+    });
+
+    //change data to JSON
+    data = JSON.parse(data)
+
+    //execute
+    await execSwingsetTransaction(
+      "wallet-action --allow-spend '" + JSON.stringify(data) + "'",
+      networkConfig,
+      from,
+      false,
+      keyring,
+    );
+
+    //update last submission time
+    await updateTable("jobs", {"last_submission_time": Date.now()/1000}, "feed")
+
+    //sleep 13 seconds to wait 2 blocks and a bit
+    await delay((Number(SEND_CHECK_INTERVAL)+1) * 1000);
+
+    //check submission for round
+    submitted = await checkSubmissionForRound(from, previousOffer, round)
+
+    //check if in submission
+    inSubmission = await checkIfInSubmission(feed)
```

Comment:
I'm not getting this, we check for submission and if we need to submit, we submit a tx and wait for it to be mined. There is no need to check if a tx is 'inSubmission'. An I missing something?

Reply:
This is done to avoid sending multiple price updates if the first price submission cannot be confirmed yet. By having inSubmission, we are able to know whether the middleware is waiting for a submission to be confirmed and that there is no need for another submission yet.